libsignal-protocol/src/raw_ptr.rs

```diff
@@ -1,62 +1,97 @@
+use std::ptr::NonNull;
+
 /// A wrapper around `libsignal-protocol`'s native reference counted pointers.
 #[derive(Debug)]
-pub struct Raw<T: SignalType>(*mut T);
+pub struct Raw<T: IsA<sys::signal_type_base>>(NonNull<T>);
 
-impl<T: SignalType> Raw<T> {
-    /// Create a new [`Raw<T>`] from an owned pointer (doesn't affect reference
+impl<T: IsA<sys::signal_type_base>> Raw<T> {
+    /// Create a new [`Raw`] from an owned pointer (doesn't affect reference
     /// count).
-    pub fn from_ptr(raw: *mut T) -> Raw<T> { Raw(raw) }
-
-    /// Create a new [`Raw<T>`] after bumping the reference count.
-    pub fn copied_from(raw: *mut T) -> Raw<T> {
-        unsafe {
-            sys::signal_type_ref(SignalType::as_signal_base(raw));
-        }
+    ///
+    /// # Safety
+    ///
+    /// This assumes `raw` points to a valid, non-null instance of `T`.
+    pub unsafe fn from_ptr(raw: *mut T) -> Raw<T> {
+        debug_assert!(!raw.is_null());
+        Raw(NonNull::new_unchecked(raw))
```

Comment:
Isn't that dangerous 😆

Reply:
Well we *do* specify it must point to a valid, non-null instance of `T` as part of the `# Safety` documentation, so the onus is on the caller to make sure the pointer isn't null. 

The `debug_assert!()` is just there to help keep us honest and detect invalid usage of `unsafe`.